docs: Better document the DMA API and expectations

teburd opened the pull request:
The DMA API has several expectations for drivers and callers that were underdocumented or undocumented. Better clarify the driver expectations and caller expectations.

The DMA API from the caller side is not a portable API and really cannot be as each DMA has unique properties and expectations of memory, peripheral interaction, and features. The API in effect provides a union of all useful DMA functionality drivers have needed in the tree. It can still be a good abstraction, with care, for peripheral devices for vendors or projects where the DMA IP might be very similar but have slight variances. It does however provide a way to configure and transition channel states. In most drivers today this can be done from any call context. This PR solidifies that this is meant to be possible in any call context and bugs have been filed for drivers that are not compliant.

From the driver implementation side expectations around synchronization, state transitions, and memory management for transfer descriptors is now described in documentation rather than solely from me in github review comments.

Commit:
docs: Better document the DMA API and expectations

The DMA API has several expectations for drivers and callers that were
underdocumented or undocumented. Better clarify the driver expectations
and caller expectations.

The DMA API from the caller side is not a portable API and really cannot
be as each DMA has unique properties and expectations of memory,
peripheral interaction, and features. The API in effect provides a union
of all useful DMA functionality drivers have needed in the tree. It can
still be a good abstraction, with care, for peripheral devices for
vendors where the DMA IP might be very similar but have slight
variances.

From the driver implementation side expectations around synchronization,
state transitions, and memory management for transfer descriptors is now
described in documentation rather than solely from me in github review
comments.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -386,6 +386,8 @@ static inline int dma_reload(const struct device *dev, uint32_t channel,
  * Start is allowed on channels that have already been started and must report
  * success.
  *
+ * @funcprops \isr_ok
```
👍

kartben commented on doc/hardware/peripherals/dma/index.rst:
```diff
@@ -0,0 +1,61 @@
+Direct Memory Access (Controller) is a commonly provided co-processor of sorts
+that can typically offload transferring data to and from peripherals and memory.
+
+The DMA API in Zephyr is not necessarily a portable API across devices but
+instead a union of common DMA features. This API can still provide a reasonable
+abstraction given some care for peripheral drivers.
```
I feel like the wording  can be improved but feel free to ignore if this doesn't sound any better to you or if it changes the meaning too much.

```suggestion
The DMA API in Zephyr does not guarantee portability across different devices.
Nevertheless, and with appropriate consideration for peripheral drivers, it can
offer a substantial level of abstraction for common DMA operations.
```

kartben replied:
btw it feels like your commit message (2nd paragraph specifically) might contain wording that's better that what ended up being captured in this paragraph, so maybe have another pass?

kartben reviewed (changes requested):
a first quick-ish pass -- thanks for this work @teburd!

teburd commented on doc/hardware/peripherals/dma/images/dma_fsm.svg:
oh thats much nicer, will change it, thanks!

kartben commented on doc/hardware/peripherals/dma.rst:
Guess you won't need to rename if you don't have the SVG anymore right?

kartben commented on doc/hardware/peripherals/dma.rst:
```diff
@@ -6,8 +6,74 @@ Direct Memory Access (DMA)
 
 API Reference
-*************
+**************
```
unneeded change.

kartben commented on doc/hardware/peripherals/dma.rst:
```diff
@@ -6,8 +6,74 @@ Direct Memory Access (DMA)
+       edge [fontname=Courier];
+       init [shape=point];
+
+       CONFIGURED [label=Configured,shape=box,fillcolor=red];
```
```suggestion
       CONFIGURED [label=Configured,shape=box];
```

Sorry, that was a leftover test 😢

kartben reviewed (changes requested):
damn, sorry for the sneaky "typo" that was hiding in my graphviz code...
looks good otherwise, thank you!

Commit:
dma: Move struct member doc comments to fields

To better cover the struct fields of the DMA API in doxygen the fields
are now individually documented rather than documenting them ad-hoc in
the struct header doc comment.

A best attempt at marking fields that are HW specific has been done as
well. That means almost all of them.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -78,65 +78,73 @@ enum dma_attribute_type {
+	/**
+	 * Destination address adjustment
+	 *
+	 * 00 increment
+	 * 01 decrement
+	 * 10 no change
```
```suggestion
	 * - 00 increment
	 * - 01 decrement
	 * - 10 no change
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -78,65 +78,73 @@ enum dma_attribute_type {
+	/**
+	 * Transfer flow control mode
+	 *
+	 * 0 source request service upon data availability
+	 * 1 source request postponed until destination request happens
```
```suggestion
	 * - 0 source request service upon data availability
	 * - 1 source request postponed until destination request happens
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -162,102 +170,113 @@ typedef void (*dma_callback_t)(const struct device *dev, void *user_data,
+	/**
+	 * Direction the transfers are occurring
+	 *
+	 * 000-memory to memory,
+	 * 001-memory to peripheral,
+	 * 010-peripheral to memory,
+	 * 011-peripheral to peripheral,
+	 * 100-host to memory
+	 * 101-memory to host
+	 * others-hardware specific
```
```suggestion
	 * - 000-memory to memory,
	 * - 001-memory to peripheral,
	 * - 010-peripheral to memory,
	 * - 011-peripheral to peripheral,
	 * - 100-host to memory
	 * - 101-memory to host
	 * - others-hardware specific
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -162,102 +170,113 @@ typedef void (*dma_callback_t)(const struct device *dev, void *user_data,
+	/**
+	 * Completion callback enable
+	 *
+	 * 0-callback invoked at transfer list completion only
+	 * 1-callback invoked at completion of each block
```
```suggestion
	 * - 0-callback invoked at transfer list completion only
	 * - 1-callback invoked at completion of each block
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -162,102 +170,113 @@ typedef void (*dma_callback_t)(const struct device *dev, void *user_data,
+	/**
+	 * Error callback enable
+	 *
+	 * 0-error callback enabled
+	 * 1-error callback disabled
```
```suggestion
	 * - 0-error callback enabled
	 * - 1-error callback disabled
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -162,102 +170,113 @@ typedef void (*dma_callback_t)(const struct device *dev, void *user_data,
+	/**
+	 * Source handshake, HW specific
+	 *
+	 * 0-HW
+	 * 1-SW
```
```suggestion
	 * - 0-HW
	 * - 1-SW
```

kartben commented on include/zephyr/drivers/dma.h:
```diff
@@ -162,102 +170,113 @@ typedef void (*dma_callback_t)(const struct device *dev, void *user_data,
+	/**
+	 * Destination handshake, HW specific
+	 *
+	 * 0-HW
+	 * 1-SW
```
```suggestion
	 * - 0-HW
	 * - 1-SW
```

kartben reviewed (changes requested):
small issue with lists that need proper bullets to not show up on a single line

bonus point for fixing L153 so that the bullet list starts on a new line and first item gets properly picked up.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/ea9bdf24-2c9c-4cad-8f34-cadc14aa69b9)
